packages/foundry/contracts/test/other/tether/TetherJoin.t.sol

```diff
@@ -0,0 +1,285 @@
+
+    function testJoinAuth() public {
+        vm.expectRevert("Access denied");
+        vm.prank(user);
+        join.join(user, unit);
+    }
+
+    function testExitAuth() public {
+        vm.expectRevert("Access denied");
+        vm.prank(user);
+        join.exit(user, unit);
+    }
+
+    function testRetrieveAuth() public {
+        vm.expectRevert("Access denied");
+        vm.prank(user);
+        join.retrieve(otherToken, other);
+    }
+    
+    function testJoinPull() public {
+        track("userBalance", tether.balanceOf(user));
+        track("storedBalance", join.storedBalance());
+        track("joinBalance", tether.balanceOf(address(join)));
+
+        vm.prank(user);
+        tether.approve(address(join), unit);
+        vm.prank(ladle);
+        join.join(user, unit);
+
+        assertTrackMinusEq("userBalance", unit, tether.balanceOf(user));
+        assertTrackPlusEq("storedBalance", unit, join.storedBalance());
+        assertTrackPlusEq("joinBalance", unit, tether.balanceOf(address(join)));
+    }
+
+    function testJoinPush() public {
+        track("userBalance", tether.balanceOf(user));
+        track("storedBalance", join.storedBalance());
+        track("joinBalance", tether.balanceOf(address(join)));
+
+        vm.prank(user);
+        tether.transfer(address(join), unit);
+        vm.prank(ladle);
+        join.join(user, unit);
+
+        assertTrackMinusEq("userBalance", unit, tether.balanceOf(user));
+        assertTrackPlusEq("storedBalance", unit, join.storedBalance());
+        assertTrackPlusEq("joinBalance", unit, tether.balanceOf(address(join)));
+    }
+
+    function testJoinCombine() public {
+        track("userBalance", tether.balanceOf(user));
+        track("storedBalance", join.storedBalance());
+        track("joinBalance", tether.balanceOf(address(join)));
+
+        vm.prank(user);
+        tether.approve(address(join), unit / 2);
+        vm.prank(user);
+        tether.transfer(address(join), unit / 2);
+        vm.prank(ladle);
+        join.join(user, unit);
+
+        assertTrackMinusEq("userBalance", unit, tether.balanceOf(user));
+        assertTrackPlusEq("storedBalance", unit, join.storedBalance());
+        assertTrackPlusEq("joinBalance", unit, tether.balanceOf(address(join)));
+    }
+}
+
+abstract contract WithFees is Deployed {
+    function setUp() public override virtual {
+        super.setUp();
+
+        // enable fees
+        vm.prank(tetherMultiSig);
+        tether.setParams(19, 49);    // maximum
+    }
+}
+
+contract WithFeesTest is WithFees {
+    function testJoinPullWithFees() public {
+        track("userBalance", tether.balanceOf(user));
+        track("storedBalance", join.storedBalance());
+        track("joinBalance", tether.balanceOf(address(join)));
+
+        uint256 units = unit * 5;
+        uint256 fee = tether.basisPointsRate() * 100;
+        uint256 feeAdjustedUnits = units * unit / (unit - fee); // scale up first so precision isn't lost
```

Comment:
Here you are subtracting the fee. Below the `tether.approve(address(join), feeAdjustedUnits);` should be for `units`, which is the amount that is expected to leave the user's address.